### PR TITLE
Fixes #36273 - Use proper permission for editing Ansible variable

### PR DESCRIPTION
**Pull request by pmoravec**
Requires:
* https://github.com/theforeman/foreman/pull/9803

### CODE REVIEW

**Comment by nofaralfasi** [comment]
Hey @pmoravec, I agree with the change you made in this PR. However, the fix only applies to the API. We also need to address the problem in the web UI. 
Furthermore, I noticed that there are no tests for this functionality. Could you please add them?

**Comment by adamruzicka** [comment]
> We also need to address the problem in the web UI.

That's true, there are two other places where arguably wrong permission (`edit_external_parameters`) is used.

>  Could you please add them?

I'd be willing to let this slide. We can reason whether the permission used is the right one or not, but the whole authorization framework is thoroughly tested in Foreman itself.

**Comment by pmoravec** [comment]
I see. Sadly I dont have bandwidth to complete the fix as requested - please let somebody else adds a fix for WebUI "variant" and add some tests.

File: lib/foreman_ansible/register.rb [hunk 1]
```diff
@@ -137,6 +137,12 @@
                  :'api/v2/ansible_variables' => [:create]
                },
                :resource_type => 'AnsibleVariable'
+    permission :view_lookup_values,
```
Comment (stejskalleos) [inline lib/foreman_ansible/register.rb:140]: Permission `:view_ansible_roles` already got `{ :lookup_values => [:index] }`

File: lib/foreman_ansible/register.rb [hunk 2]
```diff
@@ -137,6 +137,12 @@
...
+               { :lookup_values => [:index] },
+               :resource_type => 'LookupValue'
+    permission :create_lookup_values,
```
Comment (stejskalleos) [inline lib/foreman_ansible/register.rb:143]: Permission `:create_ansible_variables` already have `:lookup_values => [:create]`

**Review by stejskalleos — CHANGES_REQUESTED** [review]

File: lib/foreman_ansible/register.rb (hunk 1 above)
Comment (nofaralfasi) [reply lib/foreman_ansible/register.rb:140]: Since we don't even have permissions for the `lookup_values` resource, non-admin users can't access it (see https://github.com/theforeman/foreman/blob/develop/app/services/authorizer.rb#L34 and https://github.com/theforeman/foreman/blob/develop/app/graphql/mutations/base_mutation.rb#L31). 

You can refer to the output log below, which demonstrates the absence of filters related to the permission for the LookupValue class:
`no filters found for given permission`
`checking permission  for class LookupValue`
`organization_ids: []`
`location_ids: []`

Let me know if you need more details.

**Comment by ofedoren** [comment]
Agree on that it should be using `:*_ansible_variables` here, since as noted @stejskalleos it somehow defined to use `lookup_values` permissions as well: https://github.com/theforeman/foreman_ansible/blob/5b44890713842fd3c5f2646590ecb18265f52ebf/lib/foreman_ansible/register.rb#L119-L124.

The only thing that bothers me, I didn't find any defined permissions on LookupValue model in Foreman itself :/

**Comment by nofaralfasi** [comment]
> The only thing that bothers me, I didn't find any defined permissions on LookupValue model in Foreman itself :/

I see the LookupValue is defined in foreman core. Would it solve the issue if I add the permissions for the LookupValue resource there?

**Comment by ofedoren** [comment]
> Would it solve the issue if I add the permissions for the LookupValue resource there?

It should :shrug:

**Comment by nofaralfasi** [comment]
How to test:
1. Create a non-admin user, with `'Ansible Roles Manager'` and `View hosts` permissions.
2. UI - Login as the new user, and try to create, edit and delete Ansible variables from the host details page.
3. API - Test the following requests using the new user credentials: `POST /ansible/api/ansible_override_values`, `DELETE /ansible/api/ansible_override_values/:id`.

Note:
* Even if the user doesn't have the `Edit Host` permission, he will still be able to modify Ansible variables from the host details page. This functionality existed prior to this PR's changes, and if we want to change that behaviour, it would require disabling the corresponding option within the `Specify Matchers` section on the ansible_variable edit page.

**Comment by nofaralfasi** [comment]
The tests are not passing due to the dependency on the merge of https://github.com/theforeman/foreman/pull/9803.

**Review by ofedoren — APPROVED** [review]
Thanks, @nofaralfasi and @pmoravec, can be merged once https://github.com/theforeman/foreman/pull/9803 is in.

File: lib/foreman_ansible/register.rb [hunk 3]
```diff
@@ -160,14 +160,19 @@
                { :'api/v2/ansible_inventories' => [:schedule] }
     permission :import_ansible_playbooks,
                { :'api/v2/ansible_playbooks' => [:sync, :fetch] }
+    permission :view_lookup_values, {}, :resource_type => 'LookupValue'
+    permission :create_lookup_values, {}, :resource_type => 'LookupValue'
+    permission :edit_lookup_values, {}, :resource_type => 'LookupValue'
+    permission :destroy_lookup_values, {}, :resource_type => 'LookupValue'
```
Comment (ofedoren) [inline lib/foreman_ansible/register.rb:166]: Why is this needed? I mean, I've removed those lines and it still works...

Comment (nofaralfasi) [reply lib/foreman_ansible/register.rb:166]: Did you first remove these permissions from the DB? 
From the rails console: `Permission.where("name LIKE '%_lookup_values%'").destroy_all` 
And then re-run the rails server to test it.

Comment (ofedoren) [reply lib/foreman_ansible/register.rb:166]: I've done this and it looks... not quite good.

I've destroyed permissions, but we still need to create them, to use here. Thus, I needed to run `db:seed` to re-create them (we have defined them in core), but due to Ansible plugin needing these perms to create ARM role, it failed. Thus, after I removed Ansible plugin from loading, I've re-created the permissions, then enabled the plugin and run the server. I was able to change variable values even without this block.

I mean, we already create these perms in the core, why would we try to re-create them here? Seems weird. I've checked what we do with e.g. `create_job_invocations`, since this permission is used for ARM role as well: it seems REX plugin creates it and Ansible plugin then simply uses it to define a role. As you can see, Ansible plugin doesn't call `permission :create_job_invocations`, which is right.

Since I'm not so well familiar with the internals of permission logic, @adamruzicka, maybe you can help us out?

Comment (nofaralfasi) [reply lib/foreman_ansible/register.rb:166]: I forgot I need to make a change in the `db/seeds.d/020-permissions_list.rb` file for the `db:seed` command to include that file. That's why it didn't work as expected for me.
Therefore, as long as we run the `db:seed` command, there's no requirement to additionally include these permissions in the `register.rb` file, right?

Comment (ofedoren) [reply lib/foreman_ansible/register.rb:166]: :shrug: that's what I hoped for...

Comment (ofedoren) [reply lib/foreman_ansible/register.rb]: But the test failures now show what I was facing as well :/ Not sure why it complains about perms that are being created in DB by Foreman itself, isn't here some issues with load ordering?
